includes/admin/class-sensei-plugins-installation.php

```diff
@@ -0,0 +1,421 @@
+<?php
+/**
+ * File containing Sensei_Plugins_Installation class.
+ *
+ * @package Sensei\Admin
+ * @since 3.1.0
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Sensei Plugins Installation Class.
+ *
+ * @since 3.1.0
+ */
+class Sensei_Plugins_Installation {
+	const INSTALLING_PLUGINS_TRANSIENT = 'sensei_installing_plugins';
+
+	/**
+	 * Actions to be executed after the HTTP response has completed
+	 *
+	 * @var array
+	 */
+	private $deferred_actions = [];
+
+	/**
+	 * Instance of singleton.
+	 *
+	 * @var self
+	 */
+	private static $instance;
+
+	/**
+	 * Sensei_Plugins_Installation constructor.
+	 *
+	 * It's private to prevent other instances from being created outside of `Sensei_Plugins_Installation::instance()`.
+	 */
+	private function __construct() {
+		add_action( 'shutdown', [ $this, 'run_deferred_actions' ] );
+	}
+
+	/**
+	 * Fetches the instance of the class.
+	 *
+	 * @return self
+	 */
+	public static function instance() {
+		if ( ! self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Function called after the HTTP request is finished, so it's executed without the client having to wait for it.
+	 *
+	 * @access private
+	 */
+	public function run_deferred_actions() {
+		if ( empty( $this->deferred_actions ) ) {
+			return;
+		}
+
+		$this->close_http_connection();
+		foreach ( $this->deferred_actions as $action ) {
+			$action['func']( ...$action['args'] );
+		}
+	}
+
+	/**
+	 * Wrapper for set_time_limit to see if it is enabled.
+	 *
+	 * @param int $limit Time limit.
+	 */
+	private function set_time_limit( $limit = 0 ) {
+		if ( function_exists( 'set_time_limit' ) && false === strpos( ini_get( 'disable_functions' ), 'set_time_limit' ) && ! ini_get( 'safe_mode' ) ) { // phpcs:ignore PHPCompatibility.IniDirectives.RemovedIniDirectives.safe_modeDeprecatedRemoved
+			@set_time_limit( $limit ); // phpcs:ignore
+		}
+	}
+
+	/**
+	 * Finishes replying to the client, but keeps the process running for further (async) code execution.
+	 *
+	 * @see https://core.trac.wordpress.org/ticket/41358
+	 */
+	private function close_http_connection() {
```

Comment:
Are we confident that this approach reliably works on all servers?

Reply:
I tested in my dev env and in the Jurassic Ninja. But what makes me confident about the approach is that WC uses that. And considering that there is a lot of users using WC, I think it's very tested =)